packages/diff-sequence/diff.js

```diff
@@ -1,287 +1,268 @@
+              observer.changed && observer.changed(newDoc._id, fields);
+            }
+            observer.movedBefore && observer.movedBefore(newDoc._id, groupId);
+          }
+        }
+        if (groupId) {
+          newDoc = new_results[endOfGroup];
+          oldDoc = old_results[old_index_of_id[newDoc._id]];
+          projectedNew = projectionFn(newDoc);
+          projectedOld = projectionFn(oldDoc);
+          fields = DiffSequence.makeChangedFields(projectedNew, projectedOld);
+          if (!isObjEmpty(fields)) {
+            observer.changed && observer.changed(newDoc._id, fields);
+          }
+        }
+        startOfGroup = endOfGroup+1;
       }
-    });
-  }
-};
-
-
-DiffSequence.makeChangedFields = function (newDoc, oldDoc) {
-  var fields = {};
-  DiffSequence.diffObjects(oldDoc, newDoc, {
-    leftOnly: function (key, value) {
-      fields[key] = undefined;
     },
-    rightOnly: function (key, value) {
-      fields[key] = value;
+    diffObjects: (left, right, callbacks) => {
+      Object.keys(left).forEach(key => {
+        const leftValue = left[key];
+        if (hasOwn.call(right, key))
+          callbacks.both && callbacks.both(key, leftValue, right[key]);
+        else
+          callbacks.leftOnly && callbacks.leftOnly(key, leftValue);
+      });
+
+      if (callbacks.rightOnly)
+        Object.keys(right).forEach(key => {
+          const rightValue = right[key];
+          if (! hasOwn.call(left, key)) {
+            callbacks.rightOnly(key, rightValue);
+          }
+        });
     },
-    both: function (key, leftValue, rightValue) {
-      if (!EJSON.equals(leftValue, rightValue))
-        fields[key] = rightValue;
-    }
-  });
-  return fields;
-};
-
-DiffSequence.applyChanges = function (doc, changeFields) {
-  Object.keys(changeFields).forEach(key => {
-    const value = changeFields[key];
-    if (typeof value === "undefined") {
-      delete doc[key];
-    } else {
-      doc[key] = value;
+    diffMaps: (left, right, callbacks) => {
+      left.forEach(function (leftValue, key) {
+        if (right.has(key))
+          callbacks.both && callbacks.both(key, leftValue, right.get(key));
+        else
+          callbacks.leftOnly && callbacks.leftOnly(key, leftValue);
+      });
+      if (callbacks.rightOnly)
+        right.forEach(function (rightValue, key) {
+          if (!left.has(key)){
+            callbacks.rightOnly(key, rightValue);
+          }
+        });
+    },
+    makeChangedFields: (newDoc, oldDoc) => {
+      const fields = new Map();
+      DiffSequence.diffObjects(oldDoc, newDoc, {
+        leftOnly: function (key, value) {
+          fields.set(key, undefined)
+        },
+        rightOnly: function (key, value) {
+          fields.set(key, value);
+        },
+        both: function (key, leftValue, rightValue) {
+          if (!EJSON.equals(leftValue, rightValue))
+            fields.set(key, rightValue);
+        }
+      });
+      return Object.fromEntries(fields);
```

Comment:
Is this change performance-related? I can’t imagine how creating a `Map` instance and then transforming it into an object may be faster than creating the object straight away.